Improve the Clojure Package README to Make it Easier to Get Started (#12881)
apache/incubator-mxnet: contrib/clojure-package/README.md

## Current State and Plans

Help is needed testing and generally making the package better. A list of the pacakge status and contribution needs can be found here [Clojure Package Contribution Needs](https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs). Please get involved :)

## Getting Started

> The Clojure MXNet framework consists of a core C library, a Scala Api that talks to it through JNI bindings, and finally a Clojure wrapper around the Scala api.

Reviewer: Scala Api

Since there is a native code involved in the framework, what OS you are running matters.

The following systems are supported:

- OSX cpu
- Linux cpu
- Linux gpu

There are three ways of getting started:

> * Use the Clojure jars with the native dependencies baked in. This the easiest way to get going.

Reviewer 2: I humbly suggest section links:

```suggestion
* [Use the Clojure jars with the native dependencies baked in](#getting-started-with-the-clojure-jars-on-maven). This the easiest way to get going.
* [Checkout the MXNet project from a release tag using the Scala jars with native deps](#getting-started-with-mxnet-project-with-the-prebuilt-scala-jars). This is also a pretty easy way to get started.
* [Build from the MXNet project master](#getting-started-with-mxnet-project-with-building-from-source). This option can be used to build the whole project yourself.
```

(Note: I used GitHub's new Suggestion feature here, but it appears that it only _replaces_ the line I made the comment on, and the other two bullet points are new text rather than replacing the following lines.)

> * Checkout the MXNet project from a release tag using the Scala jars with native deps. This is also a pretty easy way to get started.

Reviewer: nit: deps refers to dependencies?

* Build from the MXNet project master. This option can be used to build the whole project yourself.

> ### Getting Started with the Clojure Jars on Maven

Reviewer: (###Getting-Started-with-the-Clojure-Jars-on-Maven) to create an anchor for this header.

Refer - [Github Markdown Heading Anchor](https://gist.github.com/asabaylus/3071099)

This option is one of the fastest. If you are looking to use MXNet in your own Clojure project, this is your best bet.

The Clojure MXNet jars are found on [maven.org](https://search.maven.org/search?q=clojure%20mxnet).

- Create your own project with `lein new my-mxnet`
- Edit your `project.clj` based on your system with the desired jar from maven.

   - clojure-mxnet-linux-gpu
   - clojure-mxnet-linux-cpu

>    - clojure-mxnet-osx-cpu

Reviewer 2: Listifying this is good. I find copy-pastable dependencies (like in the current README) helpful:

>`[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-gpu "1.3.0"]` or `[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "1.3.0"]` or `[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.3.0"]`

Perhaps:

 - `[org.apache.mxnet.contrib.clojure/clojure-mxnet-linux-gpu "1.3.0"]` 
 - `[org.apache.mxnet.contrib.clojure/clojure-mxnet-linux-cpu "1.3.0"]`
 - `[org.apache.mxnet.contrib.clojure/clojure-mxnet-osx-cpu "1.3.0"]`

?

If the specific versions or Maven coordinates are unwanted due to change then I'd advocate for an explanatory sentence, perhaps with a link to https://search.maven.org/search?q=g:org.apache.mxnet.contrib.clojure

* [Build from the MXNet project master](#getting-started-with-mxnet-project-with-building-from-source). This option can be used to build the whole project yourself.

### Getting Started with the Clojure Jars on Maven

This option is one of the fastest. If you are looking to use MXNet in your own Clojure project, this is your best bet.

The Clojure MXNet jars are found on [maven.org](https://search.maven.org/search?q=clojure%20mxnet).

- Create your own project with `lein new my-mxnet`
- Edit your `project.clj` based on your system with the desired jar from maven. Replace `x.x.x` with the latest version on [maven](https://search.maven.org/search?q=g:org.apache.mxnet.contrib.clojure).

   - [org.apache.mxnet.contrib.clojure/clojure-mxnet-linux-gpu "x.x.x"]
   - [org.apache.mxnet.contrib.clojure/clojure-mxnet-linux-cpu "x.x.x"]
   - [org.apache.mxnet.contrib.clojure/clojure-mxnet-osx-cpu "x.x.0"]

You also might need the following dependencies installed depending on your system because of the native dependencies assumed in the clojure-mxnet jars.

*For OSX you will need:*

> 

Reviewer 3: I think we also need `brew install wget` as wget isn't installed by default on osx. 

```suggestion
`brew install wget`
```

Author: Nice! Thanks @susanluke

sudo apt install libopencv-imgcodecs3.4
```

> You may also need

Reviewer: By may, do you mean it is optional to install them. If so is it because of the Clojure API?

sudo pacman -U cuda-9.0.176-4-x86_64.pkg.tar.xz
```

At this point you should be able to run your own example like this [NDArray Tutorial](https://github.com/apache/incubator-mxnet/blob/master/contrib/clojure-package/examples/tutorial/src/tutorial/ndarray.clj)

### Getting Started with MXNet project with the Prebuilt Scala Jars

This option is also fast. It doesn't require you to build the native C library or the Scala jars, it gets them from [Maven](https://search.maven.org/search?q=g:org.apache.mxnet) as well. Use this if you want to clone the repo and be able to run the tests and examples in the Clojure package.

- `git clone --recursive https://github.com/apache/incubator-mxnet.git ~/mxnet`
- `cd mxnet`
- `git tag —list` (find the tag that corresponds the version of the latest Scala jar)
- `git checkout tags/<tag_name> -b <branch_name>`
- `cd contrib/clojure`
- edit `project.clj` to include your Scala jar from maven. It should match your system. Example `[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "x.y.z”]`

> - run `lein test`. All the tests should run without an error

Author: There is a need help section now in the doc that can point them where to go

> - At this point you can do `lein install` to build and install the Clojure jar locally. Now, you can run the examples by doing `cd examples/imclassification` and then `lein run`  or `lein run :cpu 2` (for 2 cpus) (for gpu `lein run :gpu`)

Reviewer 4: Is there `lein run :gpu 8` for eight GPUs?

### Getting Started with MXNet project with Building from Source

This option will build the C lib locally and then the Scala jars as well. You can use this option when you are developing changes for the project.

* For your dependencies look at https://mxnet.incubator.apache.org/install/osx_setup.html or https://mx	net.incubator.apache.org/install/ubuntu_setup.html
	about _Prepare Environment for GPU Installation_
	and	 _Install MXNet dependencies_
* Also, ensure you have JDK 8 on your system. Later versions may produce cryptic build errors mentioning `scala.reflect.internal.MissingRequirementError`. 
* Checkout the latest SHA from the main package:
* `git clone --recursive https://github.com/apache/incubator-mxnet.git ~/mxnet`
* `cd ~/mxnet`
* `git tag —list` (find the tag that corresponds the version of the latest Scala jar)
* `git checkout tags/<tag_name> -b <branch_name>` To checkout a particular release
* `git submodule update --init --recursive` - To updated all the submodules to the commit

>   * Sometimes it is useful to use this script to clean hard, (remove any extra files and reset everything), using this [gist](https://gist.github.com/nick	toumpelis/11214362)

Reviewer 4: That url looks weird. Why the spaces?

> * Go here to do the base package installation https://mxnet.incubator.apache.org/install/index.html.

Reviewer 4: That link kinda dumps them to really generic installation info... what about this[ build from source page](https://github.com/apache/incubator-mxnet/blob/master/docs/install/build_from_source.md)?
BTW, I've updated that page with [this PR](https://github.com/apache/incubator-mxnet/pull/12962) which may or may not be merged by the time you read this.

Author: Awesome  - it's incoporated

>   * This will create an `lib/libmxnet.so` file that is the c api we need for the next step.

Reviewer 4: ```suggestion
  * This will create an `lib/libmxnet.so` file that MXNet shared library that we need for the next step.
```
There's debate about how much to expose re: the C API since it really isn't officially supported. So maybe not shine a light on it...

* Run `make scalapkg` This will create the Scala jars.

> * Run `make scalainstall`. This will install them in your local maven.

Reviewer 4: ```suggestion
* Run `make scalainstall`. This will install them in your local Maven.
```

> * `cd contrib/clojure` and edit the project.clj to include the Scala jar that was just created and installed in your maven. Example `[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.3.0-SNAPSHOT"]`.

Reviewer 4: ```suggestion
* `cd contrib/clojure` and edit the project.clj to include the Scala jar that was just created and installed in your Maven. For example, for macOS CPU: `[org.apache.mxnet/mxnet-full_2.11-osx-x86_64-cpu "1.3.0-SNAPSHOT"]`.
```

* run `lein test`. All the tests should run without an error
*  At this point you can do `lein install` to build and install the clojure jar locally
*  Now, you can run the examples by doing `cd examples/imclassification` and then `lein run`  or `lein run :cpu 2` (for 2 cpus) (for gpu `lein run :gpu`)

## Docker Files

There are Dockerfiles available as well.

> - [Community Provided by Magnet](https://hub.docker.com/u/magnetcoop/)

Reviewer 4: ```suggestion
- [Community-provided by Magnet](https://hub.docker.com/u/magnetcoop/)
```

- [MXNet CI](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/Dockerfile.build.ubuntu_cpu) and the install scripts
  - [ubuntu core](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_core.sh)
  - [ubuntu scala](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_scala.sh)
  - [ubuntu clojure](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_clojure.sh)

## Need Help?

If you are having trouble getting started or have a question, feel free to reach out at:

- Clojurian Slack #mxnet channel. To join, go to [http://clojurians.net/](http://clojurians.net/).

> - Apache Slack #mxnet and #mxnet-scala channel (To join this slack send an email to dev@mxnet.apache.org)

Reviewer 4: ```suggestion
- Apache Slack #mxnet and #mxnet-scala channel. To join this slack send an email to dev@mxnet.apache.org.
```

> - Create an Issue on [https://github.com/apache/incubator-mxnet/issues](https://github.com/apache/incubator-mxnet/issues)

Reviewer 4: ```suggestion
- Create an Issue on [https://github.com/apache/incubator-mxnet/issues](https://github.com/apache/incubator-mxnet/issues).
```

sudo pacman -U cuda-9.0.176-4-x86_64.pkg.tar.xz
```

At this point you should be able to run your own example like this [NDArray Tutorial](https://github.com/apache/incubator-mxnet/blob/master/contrib/clojure-package/examples/tutorial/src/tutorial/ndarray.clj)

### Getting Started with MXNet project with the Prebuilt Scala Jars

> This option is also fast. It doesn't require you to build the native C library or the Scala jars, it gets them from [Maven](https://search.maven.org/search?q=g:org.apache.mxnet) as well. 

Reviewer 2: It would be slightly redundant, but I think maybe a sentence here similar to the one under *Getting Started with the Clojure Jars on Maven* ("If you are looking to use MXNet in your own Clojure project, this is your best bet.") would provide confusion-alleviating context.

Perhaps "Use this if you want your Clojure project to use a version of MXNet that hasn't yet landed on Maven."—or perhaps some other use case unknown to me is more relevant.

- `git clone --recursive https://github.com/apache/incubator-mxnet.git ~/mxnet`
- `cd mxnet`

> - `git tag —list` (find the tag that corresponds the the version of the latest Scala jar)

Reviewer: typo

- `git checkout tags/<tag_name> -b <branch_name>`
- `cd contrib/clojure`
- edit `project.clj` to include your Scala jar from maven. It should match your system. Example `[org.apache.mxnet/mxnet-full_2.11-linux-x86_64-cpu "x.y.z”]`
- run `lein test`. All the tests should run without an error

> - At this point you can do `lein install` to build and install the clojure jar locally. Now, you can run the examples by doing `cd examples/imclassification` and then `lein run`  or `lein run :cpu 2` (for 2 cpus) (for gpu `lein run :gpu`)

Reviewer: nit: to maintain consistency, Clojure (c capital) and jar ( j small) - throughout the document

> You also might need dependencies installed based on your system. Please see the dependency section in above in _Getting Started with the Clojure Jars on Maven_

Reviewer: You could create a corresponding link here - `[Getting Started with the Clojure Jars on Maven](###Getting-Started-with-the-Clojure-Jars-on-Maven)`

> You also might need dependencies installed based on your system. Please see the dependency section in above in _Getting Started with the Clojure Jars on Maven_

Reviewer 2: typo: "in above in"

```suggestion
You also might need dependencies installed based on your system. Please see the dependency section above in _Getting Started with the Clojure Jars on Maven_.
```

> ### Getting Started with MXNet project with Building from Source

Reviewer: Since all the previous steps were given as bullet points, we could maintain the same way for this section as well.

If you need to checkout a particular release you can do it with:

`git checkout tags/<tag_name> -b <branch_name>`

`git submodule update --init --recursive`

> Sometimes it useful to use this script to clean hard using this [gist](https://gist.github.com/nicktoumpelis/11214362)

Reviewer: it is*

Reviewer: What is mean't by "clean hard"

- run `lein test`. All the tests should run without an error
- At this point you can do `lein install` to build and install the clojure jar locally. Now, you can run the examples by doing `cd examples/imclassification` and then `lein run`  or `lein run :cpu 2` (for 2 cpus) (for gpu `lein run :gpu`)

## Docker Files

There are Dockerfiles available as well.

- [Community Provided by Magnet](https://hub.docker.com/u/magnetcoop/)
- [MXNet CI](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/Dockerfile.build.ubuntu_cpu) and the install scripts
  - [ubuntu core](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_core.sh)
  - [ubuntu scala](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_scala.sh)
  - [ubuntu clojure](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_clojure.sh)

## Need Help?

If you are having trouble getting started or have a question, feel free to reach out at:

> - Clojurian Slack #mxnet channel

Reviewer: How to join this channel?

See the Confluence page: https://cwiki.apache.org/confluence/display/MXNET/MXNet+Clojure

## Building and Deploying Jars

> The release process for deploying the Clojure jars is on the [wiki](https://cwiki.apache.org/confluence/display/MXNET/Clojure+Release+Process).

Reviewer 4: ```suggestion
The release process for deploying the Clojure jars is on the [Apache MXNet developer wiki](https://cwiki.apache.org/confluence/display/MXNET/Clojure+Release+Process).
```
